File: haskell/mergesort.hs
merge :: Ord a => [a] -> [a] -> [a]
merge [] xs = xs
merge ys [] = ys
merge xs@(x:xt) ys@(y:yt) | x <= y = x : xt `merge` ys
                          | otherwise = y : xs `merge` yt

merge' :: (a -> a -> Bool) -> [a] -> [a] -> [a]
merge' _ xs [] = xs
merge' _ [] ys = ys
merge' p xs@(x:xt) ys@(y:yt) | p x y = x : merge' p xt ys
                             | otherwise = y : merge' p xs yt

split :: [a] -> ([a], [a])
split xs = splitAt (length xs `quot` 2) xs

-- split (x:y:zs) = let (xs, ys) = split zs in (x:xs, y:ys)
-- split [x] = ([x], [])
-- split [] = ([], [])

-- split xs = go xs xs where
--     go (x:xs) (_:_:zs) = (x:us,vs) where (us,vs) = go xs zs
--     go xs _ = ([], xs)

sort :: Ord a => [a] -> [a]
sort [] = []
sort [x] = [x]
sort xs = let (lft, rgt) = split xs
          in sort lft `merge` sort rgt

sort' :: (a -> a -> Bool) -> [a] -> [a]
sort' _ [] = []
sort' _ [x] = [x]
sort' p xs = let (lft, rgt) = split xs
             in merge' p (sort' p lft) (sort' p rgt)

main = do
    print $ sort ls
    print $ sort' (<=) ls
    where ls = [ 4, 3, 2, 1 ]

File: swift/mergesort.swift
func merge<T: Comparable>(lft: [T], rgt: [T]) -> [T] {
    func head(arr: [T]) -> T {
        return arr[0]
    }

    func tail(arr: [T]) -> [T] {
        return Array(arr[1 ..< arr.count])
    }

    if lft.isEmpty {
        return rgt
    } else if rgt.isEmpty {
        return lft
    } else if head(lft) < head(rgt) {
        return [ head(lft) ] + merge(tail(lft), rgt)
    } else {
        return [ head(rgt) ] + merge(lft, tail(rgt))
    }
}

func sort<T: Comparable>(arr: [T]) -> [T] {
    func from(lo: Int, hi: Int) -> [T] {
        return Array(arr[lo ..< hi])
    }

    if arr.count > 1 {
        let mid = arr.count / 2
        return merge(sort(from(0, mid)), sort(from(mid, arr.count)))
    }

    return arr
}

println(sort([5, 4, 3, 2, 1]))

func merge<T>(lft: [T], rgt: [T], pred: (T, T) -> Bool) -> [T] {
    func head(arr: [T]) -> T {
        return arr[0]
    }

    func tail(arr: [T]) -> [T] {
        return Array(arr[1 ..< arr.count])
    }

    if lft.isEmpty {
        return rgt
    } else if rgt.isEmpty {
        return lft
    } else if pred(head(lft), head(rgt)) {
        return [ head(lft) ] + merge(tail(lft), rgt, pred)
    } else {
        return [ head(rgt) ] + merge(lft, tail(rgt), pred)
    }    
}

func sort<T>(arr: [T], pred: (T, T) -> Bool) -> [T] {
    func from(lo: Int, hi: Int) -> [T] {
        return Array(arr[lo ..< hi])
    }

    if arr.count > 1 {
        let mid = arr.count / 2
        return merge(sort(from(0, mid), pred), sort(from(mid, arr.count), pred), pred)
    }       

    return arr
}

println(sort([1, 2, 3, 4, 5], { $0 > $1 }))

extension Array {
    func slice(min: Int, _ max: Int) -> [T] {
        return Array(self[min ..< max])
    }
    
    var head: T {
        return self[0]
    }

    var tail: [T] {
        return self.slice(1, self.count)
    }

    func split() -> ([T], [T]) {
        let mid = self.count / 2
        return (slice(0, mid), slice(mid, self.count))
    }
}

func merge_<T>(lft: [T], rgt: [T], pred: (T, T) -> Bool) -> [T] {
    if lft.isEmpty {
        return rgt
    } else if rgt.isEmpty {
        return lft
    } else if pred(lft.head, rgt.head) {
        return [ lft.head ] + merge_(lft.tail, rgt, pred)
    } else {
        return [ rgt.head ] + merge_(lft, rgt.tail, pred)
    }
}

func sort_<T>(arr: [T], pred: (T, T) -> Bool) -> [T] {
    if (arr.count > 1) {
        let (lft, rgt) = arr.split()
        return merge_(sort_(lft, pred), sort_(rgt, pred), pred)
    }

    return arr
}

println(sort_([1, 2, 3, 4, 5], { $0 > $1 }))